SCICD-504: Sanity Checks

larry-tordsen opened the pull request:
Check the media directory for new tarballs that haven't been run through process-media yet.

Ensure the media_dir is always defined.  It doesn't always need to be explicitly specified, but we should be able to determine what it is.  If it cannot be determined, it is an error.

ltordsen committed:
SCICD-504: Sanity Checks

Check the media directory for new tarballs that haven't been run through
process-media yet.

Ensure the media_dir is always defined.  It doesn't always need to be
explicitly specified, but we should be able to determine what it is.  If it
cannot be determined, it is an error.

JohnNavitskyHPE commented on lib/Activity.py:
```diff
@@ -817,7 +828,38 @@ def watch_next_wf(self, sessionid):
             outfile = os.path.join(sessions_dir, f"{wfid}-{stage}.yaml")
             with open(outfile, "w", encoding='UTF-8') as fhandle:
                 yaml.dump(cfgmaps, fhandle)
+            files = [file for file in os.listdir(self.media_dir) if "tar" in file]
```
we might want to tighten this up a bit more, looking for _specifically_ '.tar$' or '.tar.gz$'

JohnNavitskyHPE commented on lib/Activity.py:
```diff
@@ -817,7 +828,38 @@ def watch_next_wf(self, sessionid):
+
+            for product in files:
+                prodpath = os.path.join(self.media_dir, product)
+                tar_paths = self.config.connection.sudo(f"tar tvf {prodpath}").stdout.splitlines()
```
i don't think we want to do an unconstrained tar here as some of the tarballs are very large and even just doing a listing takes a long time.
```
ncn-m001:/etc/cray/upgrade/csm/fullstack-iw3 # time FOO=$(tar tzf csm-1.4.0-beta.77.tar.gz)

real	3m35.657s
user	3m16.893s
sys	0m26.180s
ncn-m001:/etc/cray/upgrade/csm/fullstack-iw3 # time FOO=$(tar tzf csm-1.4.0-beta.77.tar.gz | head -n 1)

real	0m0.011s
user	0m0.014s
sys	0m0.000s
ncn-m001:/etc/cray/upgrade/csm/fullstack-iw3 # echo $FOO
csm-1.4.0-beta.77/
ncn-m001:/etc/cray/upgrade/csm/fullstack-iw3 #
```

JohnNavitskyHPE commented on lib/Activity.py:
```diff
@@ -445,11 +467,11 @@ def dfs(data, start):
             dpath = dfs(slist, workflow)
         except:
-            """ if dfs fails it's because the workflow is still starting up, just punt until the next 1 second check"""
+            # If dfs fails it's because the workflow is still starting up, just punt until the next 1 second chec.
```
s/chec/check/

ltordsen committed:
SCICD-618: Create Did Not Return an Error

Issue an error if the '--create' flag was passed to activity without
specifying the state positional argument.

ltordsen committed:
SCICD-682: Media Dir is getting set to /opt/cray/iuf

The media directory should default to the RBD_BASE_DIR rather than
/opt/cray/iuf.

ltordsen committed:
SCICD-681: Catch a General Exception in PodLogs

In PodLogs, catch a general exception and print a debug log message.

JohnNavitskyHPE commented on lib/Activity.py:
```diff
@@ -817,7 +829,53 @@ def watch_next_wf(self, sessionid):
+                    if "iuf-product-manifest.yaml" in dir_contents:
+                        matches = [cp for cp in cfgmap_products if cp in product]
+                        if len(matches) < 1 and stage != "process-media":
+                            self.config.logger.warning(f"It looks like process-media hasn't been ran since {product} was added to the media directory")
```
How about "Tar file {tar_path} found in media directory but is not present in the list of products IUF is working with, you will need re-run starting from the process-media stage to include this product."

JohnNavitskyHPE commented on lib/Activity.py:
```diff
@@ -817,7 +829,53 @@ def watch_next_wf(self, sessionid):
+                elif tar_path:
+                    # The tarfile exists within the media directory, but it hasn't been extracted.
+                    if stage != "process-media":
+                        self.config.logger.debug(f"New tarfile found: {prodpath}.  Was process-media ran since it was added?")
```
How about "Tar file {tar_path} found in media directory but is not present in the list of products IUF is working with, you will need re-run starting from the process-media stage to include this product."
